node/server.js

```diff
@@ -5154,6 +5125,135 @@ function init_io() {
 					instance.players[NPC_prefix + npc.id] = npc;
 				}
 			} else {
+				const gMap = data.place && G.maps[data.place];
+				if (gMap && gMap.instance) {
+					// Generic instance creation and transportation
+					// data.place is the name of the instance e.g bee_dungeon
+					// data.name is the id of the instance usually a number
+
+					const instanceExists = data.name && instances[data.name] && instances[data.name].map == data.place;
+
+					if (data.name) {
+						// Player requested to enter an existing instance
+						if (!instances[data.name]) {
+							// The instance doesn't exist
+							return fail_response("transport_cant_invalid");
+						}
+					}
+
+					// Requirements for entering
+					if (gMap.enter) {
+						// Is there a place you need to be near to use the key
+						const validateRange = gMap.enter.locations && gMap.enter.locations.length > 0;
+						let inRange = !validateRange;
+
+						if (validateRange) {
+							server_log(`${gMap.enter.locations.length} locations to validate`);
+							for (const [locationsMapKey, locationType, locationIndex, range = 120] of gMap.enter.locations) {
+								if (!G.maps[locationsMapKey]) {
+									server_log(`G.maps.${locationsMapKey} does not exist`);
+									continue;
+								}
+
+								if (!G.maps[locationsMapKey][locationType]) {
+									server_log(`G.maps.${locationsMapKey}.${locationType} does not exist`);
+									continue;
+								}
+
+								if (!G.maps[locationsMapKey][locationType][locationIndex]) {
+									server_log(`G.maps.${locationsMapKey}.${locationType}.${locationIndex} does not exist`);
+									continue;
+								}
+
+								const location = G.maps[locationsMapKey][locationType][locationIndex];
+								const distanceToLocation = simple_distance(player, {
+									in: locationsMapKey,
+									map: locationsMapKey,
+									x: location[0],
+									y: location[1],
+								});
+
+								server_log(
+									`G.maps.${locationsMapKey}.${locationType}.${locationIndex} ${distanceToLocation} <= ${range}`,
+								);
+
+								if (distanceToLocation <= range) {
+									inRange = true;
+									break;
+								}
+							}
+						}
+
+						if (!inRange) {
+							return fail_response("transport_cant_reach");
+						}
+
+						// We check that the instance does not exist so the cost is not paid multiple times.
+						if (gMap.enter.items && !instanceExists) {
+							const itemsToConsume = [];
+
+							const quantityByItem = {};
+							for (const [itemKey, quantity] of gMap.enter.items) {
+								quantityByItem[itemKey] = quantity;
+							}
```

Comment:
Yes! Instead of tuples! ([Here](https://github.com/kaansoral/adventureland/pull/135/files#diff-0b0b6798b6b4abd01884f44fc740b99592d1c8e36f1602347210b909d911097fR148), for example)

Reply:
should be solved